
package chat;

enum MSG
{
	Login_Request		=	10001;
	Login_Response		=	10002;
	Logout_Request		=	10003;
	Logout_Response		=	10004;
	Keepalive_Request	=	10005;
	Keepalive_Response	=	10006;

	Get_Friends_Request	=	10007;
	Get_Friends_Response =	10008;
	Send_Message_Request	= 10009;
	Send_Message_Response	= 10010;

	Friend_Notification	= 20001;
	Message_Notification	= 20002;
	Welcome_Notification	= 20003;
}

message LoginRequest
{
	required bytes username = 1;
	optional string password = 2;
}

message LoginResponse
{
	required fixed32 ttl = 1;
}

message Friend
{
	required bytes name		=	1;
	optional bool  online	=	2;
}

message GetFriendsResponse
{
	repeated Friend		friends		=	1;
}

message SendMessageRequest
{
	optional bytes	receiver	=	1;
	required string  text		=	2;
}

message FriendNotification
{
	required bytes name		=	1;
	optional bool	online	=	2;
}

message MessageNotification
{
	required string	sender	=	1;
	required string	text	=	2;
	required string	timestamp	=	3;
}

message WelcomeNotification
{
	required  bytes text	=	1;
}

message Request
{
	optional LoginRequest	login	=	1;
	optional SendMessageRequest	send_message	=	2;
}

message Response
{
	required bool	result	=	1;
	required bool	last_response	=	2;
	optional bytes	error_describe	=	3;

	optional LoginResponse	login	= 4;
	optional GetFriendsResponse	get_friends	= 5;
}

message Notification 
{
	optional FriendNotification	friend	=	1;
	optional MessageNotification	msg	=	2;
	optional WelcomeNotification	welcome	= 3;
}

message Message 
{
	required MSG		msg_type	=	1;
	required fixed32	sequence	=	2;
	optional fixed32	session_id	=	3;

	optional Request	request		=	4;
	optional Response	response	=	5;
	optional Notification notification	=	6;
}